Derive token policy forging keys according to CIP-1855

sevanspowell opened the pull request:
# Issue Number

ADP-1044

## Overview

- Add new AddressDerivation module implementing policy key derivation described in CIP-1855 (for minting and burning).
- Add PolicyK option to the "Depth" enum (for mint/burn policy keys).
- Add minimal tests to test the key derivation.

sevanspowell committed:
Resurrect hashVerificationKey

- Resurrect hashVerificationKey (it was deleted by weeder).

sevanspowell committed:
CIP-1855: Add primitives for MintBurn policy key derivation

- Add primitives for deriving policy keys used to mint/burn assets. Implemented
  according to [CIP-1855](https://github.com/cardano-foundation/CIPs/blob/b2e9d02cb9a71ba9e754a432c78197428abf7e4c/CIP-1855/CIP-1855.md).
- Add "PolicyK" option to the "Depth" enum.
- Add simple tests for key derivation.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -167,7 +170,14 @@ import qualified Data.Text.Encoding as T
+    | RoleK
+    | AddressK
+    | ScriptK
+    | PolicyK
```
Better update all the comments about `Depth`, starting with the one just above.

sevanspowell replied:
Mmmm... not ideal. Kinda brutalizing the abstraction here, but 'ScriptK did it first and it's the path of least resistance...

sevanspowell replied:
Updated doco.

rvl replied:
Yeah if it weren't for ScriptK I would have asked you to try and find a better way of representing depth.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -492,6 +502,14 @@ deriveRewardAccount pwd rootPrv =
+hashVerificationKey
+    :: WalletKey k
+    => KeyRole
+    -> k d XPub
```
```suggestion
    -> key depth XPub
```

sevanspowell replied:
Funnily enough I wanted to do this originally but LSP was giving me errors. Seems to be fine now :+1:

sevanspowell replied:
Fixed.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/MintBurn.hs:
```diff
@@ -0,0 +1,99 @@
+import Cardano.Crypto.Wallet.Types
+    ( DerivationScheme (DerivationScheme2) )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (PolicyK, PurposeK, RootK)
```
Please just use wildcards for constructor and record imports - then it's less painful to modify the code later.
```suggestion
    ( Depth (..)
```

sevanspowell replied:
:thinking: I wonder if there's an LSP setting for this. These imports are the way they are because I clicked "LSP: import this for me".

sevanspowell replied:
Fixed.

rvl replied:
Yes LSP does the same for me, but it's so excessive ... and if you have LSP it shows where a symbol comes from in the tooltip.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/MintBurn.hs:
```diff
@@ -0,0 +1,99 @@
+  where
+    policyK = liftRawKey policyPrv
+    policyPrv = derivePolicyPrivateKey pwd (getRawKey rootPrv) policyIx
+    vkeyHash = hashVerificationKey CA.Payment (publicKey policyK)
```
Purpose field here doesn't look right ... right?

sevanspowell replied:
You mean the `CA.Payment` value? I stripped out the `Role -> KeyRole` indirection because I found that confusing when initially using the function. As someone who wasn't familiar with the wallet, `Payment | Delegation :: KeyRole` made a lot more sense than `UtxoExternal | UtxoInternal | MutableAccount :: Role`, and I figured since I was the only one using it, I might as well use the simpler option. Also it doesn't feel like there's much need for abstraction here, we're literally reaching in and manipulating bytes. But I'm not strongly opposed to putting the `Role -> KeyRole` indirection back in.

rvl replied:
Yeah that's all fine, but `KeyHash { role = Payment }` ? Actually I suppose Payment is suitable for this.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/MintBurn.hs:
```diff
@@ -0,0 +1,99 @@
+  -- ^ Index of policy script
+  -> (key 'PolicyK XPrv, KeyHash)
+  -- ^ Policy private key
+derivePolicyKey pwd rootPrv policyIx = (policyK, vkeyHash)
```
Looks good - perhaps you can think of a more different name for this function?

sevanspowell replied:
Done.

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/MintBurnSpec.hs:
```diff
@@ -0,0 +1,95 @@
+            property prop_keyDerivationFromXPrv
+        it "Policy public key hash matches private key" $
+            property prop_keyHashMatchesXPrv
+
```
Perhaps one or two golden tests that you have generated manually with cardano-cli?

rvl reviewed (approved):
Looks good :+1:

sevanspowell committed:
Update Depth commentary to reflect current usage

- Update the commentary of the "Depth" data type to better reflect it's current
  usage.

sevanspowell committed:
Add more explicit type variable parameters to "hashVerificationKey"

sevanspowell committed:
More descriptive name for "derivePolicyKey"

sevanspowell committed:
Use wildcards for constructor imports

sevanspowell committed:
Add a couple more property tests to policy key derivation

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/MintBurnSpec.hs:
```diff
@@ -0,0 +1,145 @@
+prop_keyDerivationDiffIndexDiffKey pwd masterkey policyIx1 policyIx2 =
+    key1 =/= key2
+  where
+    key1 :: XPrv
+    key2 :: XPrv
```
```suggestion
    key1, key2 :: XPrv
```
... or just leave out the type annotations.

sevanspowell committed:
Compare wallet policy key derivation to cardano-addresses

- Add some golden tests comparing wallet policy key derivation to
  cardano-addresses policy key derivation.

rvl commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/MintBurnSpec.hs:
```diff
@@ -0,0 +1,237 @@
+                -- Hash of child key 1855H/1815H/0H generated by cardano-address
+                -- CLI from test mnemonic
+            unit_comparePolicyKeys goldenTestMnemonic (Index 0x80000010)
+                "acct_xsk1rprds8krzhspy8xhkupk2270hzkexmkdga98zv5ns4x9d9kf89t4qp438a26sn62wsfldthkrvnwe6vnetuehgxyxt4hm46zsw03mknuwef0afcljmultpjjwx8zqm0ftnfkqucwvf5qr4gved6h3gnyeqcsur4p"
```
cardano-address did not produce the correct bech32 prefix for policy keys, but never mind.

rvl reviewed (approved):
Excellent, thanks @sevanspowell !
